### ISSUE TITLE
GitHub pages based blogging platform

### ISSUE THREAD

**Issue by alexcorvi** [Lamassau/future-plans#1]
# GitHub pages based blogging platform

## Short Definition

A blogging platform, that generates static HTML files and directory structure out of markdown, supports both left-to-right & right-to-left languages, multiple authors and syntax highlighting.

### Description

Creating a markdown to html static site generator is trivial when using [marked](https://github.com/chjj/marked) with Node.js. Hence, It'll be very much educational with minimal experience barrier. Moreover, the idea of having a shared blog for the group members have been circling around for a while now, and we haven't found an all-in-one solution yet.

Posting new articles is just forking the blog, adding new markdown file, and sending a pull request. The pull request will then get reviewed by peers and then merged (hence published).

### Goals

- An educational project with a minimal barrier, that anyone with JavaScript, CSS and/or HTML background can contribute to.
- A blogging platform that we all use, and we all know the inner working of. Truly and fully customizable solution, with no server-side required.

### Expected Roadmap

After creating the initial functioning platform, collaborators can send pull requests that adds:

- Tag support
- Category support
- CSS enhancements
- Author biographies and pages
- Implement few things using GitHub API (a good example is displaying the pinned repos on the biography of each author)
- And more ...

### Target Audience/Users

Mainly the group, but it should be made to be reusable for anyone who wants to create start a static blog.

### Barriers & Limitations

Barriers: Minimal

Limitations:
- Compared to medium, we won't have the like-minded community we would have if we chose medium
- Compared to ghost (and wordpress), we won't have the web interface to post and manage articles.

### Available Alternatives

- Static site generators: Jekyll and Hexo. However, this project is meant to be educational, in addition to the ability to implement any feature we can possibly think of.
- Medium: Doesn't support syntax highlighting, doesn't support right-to-left languages.
- Ghost, WordPress: requires server side.
- Currently there is a Jekyll blog being hosted on this organization created by @lnyousif but I think we can make our own for educational purposes.
- @AlKass proposed using his Django blogging project, but I doubt that many of us have a good Django background. 

### Effort

A week, maximum, to get the basic functionalities. Then as time permits we can work together and implement feature by feature.

### Possible Tech Stack

- I'd suggest node and javascript, but I really do recommend using [TypeScript](http://www.typescriptlang.org/). It will make working on the project twice as productive, reduce bugs, and avoid beginners from making common mistakes, not to mention that it has been designed to make javascript suitable for collaborative projects. For me it was the easiest language I've ever been introduced to. It's literally just JavaScript with type annotations, and all the javascript tooling we've used to and love can be used with typescript with minimal headache.
- [Disqus](https://disqus.com/) for comments.
- [Highlight.js](https://highlightjs.org/) for syntax highlighting.
- [Marked](https://github.com/chjj/marked) for markdown to HTML conversion.

----------

* I have a good background about what's being proposed.
	- Yes.
* The proposal is a fork of another project.
	- No.
* I have already started working on what's being proposed.
	- No.
* I have free time to dedicate to this project.
	- Yes.

----------

> __Note__:
> This is just a proposal, I'd be happy to be introduced to a better alternative or corrected if I'm mistaken. Any opinion will be highly appreciated.

**Comment by fadeenk** [comment]
Lets do voting via thumbs up and down on the main post.

If you approve thumbs up, if not thumbs down. Comments are welcome :)

**Comment by mujz** [comment]
You should also look at [Hugo](https://github.com/spf13/hugo)

**Comment by alexcorvi** [comment]
@AlKass @fadeenk @lnyousif 

So we have the policy to listen/discuss proposals but we don't have one to accept or dismiss any.

Here's my suggestion:

For a proposal to be accepted and become part of the organization:
1. A repository should be created outside the organization.
2. At least 3 members should contribute to the repository (which is still outside the org). This way we ensure that there's a kind of common understanding of the codebase.
3. The code must have at least met its primary goals of producing a functional product.
4. Then it must be scrutinized by other members that haven't contributed to it yet, providing feedback via issue submissions and pull requests.
5. Finally, it will be transferred to Lamassau org.

What do you think?

**Comment by lnyousif** [comment]
@AlKass  I agree with the goal but the guidelines are a bit overkill for ideas to be implemented, we can clear Repos that we feel are dead, But opening them here may help bring visibility to the work more than a separate place

**Comment by AlKass** [comment]
I'm in agreement with a combination of what both @alexcorvi  and @lnyousif have suggested. one thing I'm feeling the need to object to is #2 in @alexcorvi's suggestion. This could potentially throw off a lot of novice programmers with great ideas and not sufficient skills to implement them. We have a team of very talented people who are capable of producing a lot, and sometimes it's the idea that you need not the skill set of the person proposing it.

**Comment by alexcorvi** [comment]
@lnyousif The reason I'm suggesting that the repository should be created elsewhere is that not all contributors have permission to create repositories here. As I recall that's actually what happened with PiTrol, initially, @fadeenk created the repository in his account then it got transferred here. correct me if I'm wrong.

@AlKass After thinking about it, yeah that's true. But then at least it should be reviewed, I don't think having a repository that can not be maintained by anyone other than the original author is a good idea, right?

Anyways, what policy do you suggest? because, judging from this proposal, I don't think we'll get many downvotes. I'm pretty sure that there's someone here thinks that this proposal is stupid but didn't downvote (to avoid hurting my feelings maybe? lol).

**Comment by AlKass** [comment]
You have my agreement on that
